File: lib/entity/talk.dart
import 'package:cloud_firestore/cloud_firestore.dart';

class Talk {
  String content;
  String sender;
  DateTime createdAt;

  Talk({required this.content, required this.sender, required this.createdAt});

  factory Talk.fromSnapShot(DocumentSnapshot snapshot) {
    return Talk(
      content: snapshot.get('content'),
      sender: snapshot.get('sender'),
      createdAt: (snapshot.get('createdAt') as Timestamp).toDate(),
    );
  }

  toJson() {
    return {
      "content": content,
      "sender": sender,
      "createdAt": Timestamp.fromDate(createdAt),
    };
  }
}

File: types/talk.ts
import { Timestamp, type DocumentData, type QueryDocumentSnapshot } from "firebase/firestore";

export type Talk = {
  id: string;
  content: string;
  sender: string;
  createdAt: Date;
};

export type TalkInput = Omit<Talk, "id">;

export function fromTalkDoc(doc: QueryDocumentSnapshot<DocumentData>): Talk {
  const d = doc.data();
  return {
    id: doc.id,
    content: d.content,
    sender: d.sender,
    createdAt: (d.createdAt as Timestamp).toDate(),
  };
}

export function toTalkDoc(t: TalkInput): DocumentData {
  return {
    content: t.content,
    sender: t.sender,
    createdAt: Timestamp.fromDate(t.createdAt),
  };
}
